### PR TITLE
[#423] Notifications module

### PR DESCRIPTION
**Pull request by michal-szostak**
## What is in this PR?

Notification module with basic functionality and very basic (based on bootstrap) styles.
@abacz is assumed to style them further, but for now they should be at least presentable.

## How does it looks?

image

## Other noteworthy things

The module is a separate library, which cuts it's dependencies to the main application.

Closes: #423

### CODE REVIEW

File: s4e-web/projects/notifications/src/lib/components/general-notification/general-notification.component.html [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<div [ngClass]="{clickable: notification.clearable}"
+     (click)="activated.emit()"
+     class="notification-container ui-notification {{notification.type}}"
+     role="alert">
+<!--  <h3 *ngIf="notification.caption">{{notification.caption}}</h3>-->
+  <div class="message" [innerHtml]="notification.content"></div>
+<!--  <div *ngIf="notification.clearable" class="notification-action">-->
```
Comment (danielkryska) [inline s4e-web/projects/notifications/src/lib/components/general-notification/general-notification.component.html:7]: Do you need those comments?

File: s4e-web/projects/notifications/src/lib/state/notification.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [NotificationsModule.forRoot()],
+    });
+
+    notificationService = TestBed.get(NotificationService);
+    notificationStore = TestBed.get(NotificationStore);
+  });
+
+  it('TODO - redo tests', () => {
```
Comment (danielkryska) [inline s4e-web/projects/notifications/src/lib/state/notification.service.spec.ts:19]: What test case is it?

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,6 +20,7 @@ The most recent changes are on top, in each type of changes category.
 - Favorites products [#420] (https://github.com/cyfronet-fid/sat4envi/issues/420)
 - Favourite Products [#403](https://github.com/cyfronet-fid/sat4envi/issues/403)
 - Forms module [#370](https://github.com/cyfronet-fid/sat4envi/issues/370)
+- Notifications module [#423](https://github.com/cyfronet-fid/sat4envi/issues/423)
```
Comment (jswk) [inline CHANGELOG.md:23]: This should go to the unreleased section

File: s4e-web/package.json [hunk 4]
```diff
@@ -3,11 +3,19 @@
   "version": "0.0.2",
   "scripts": {
     "ng": "ng",
+    "start-notifications": "ng build notifications --watch",
+    "build-libraries": "ng build notifications",
+    "prestart": "node_modules/.bin/npm run build-libraries",
```
Comment (jswk) [inline s4e-web/package.json:8]: Does it have to be repeated so many times? Isn't there some way to automatically run the build-libraries when running/building the app?

File: s4e-web/projects/notifications/src/lib/components/notification-outlet/notification-outlet.component.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  constructor(private notificationQuery: NotificationQuery,
+              @Inject(NotificationAction) private notificationActions: any[],
+              private notificationService: NotificationService) { }
+
+  ngOnInit() {
+    this.notifications$ = this.notificationQuery.selectAll();
+  }
+
+  activated(id: ID, ctx?: string) {
+    console.log(this.notificationActions);
```
Comment (jswk) [inline s4e-web/projects/notifications/src/lib/components/notification-outlet/notification-outlet.component.ts:40]: Should this be logged?

Comment (michal-szostak) [reply s4e-web/projects/notifications/src/lib/components/notification-outlet/notification-outlet.component.ts:40]: right, my bad

File: s4e-web/package.json (hunk 4 above)
Comment (michal-szostak) [reply s4e-web/package.json:8]: unfortunately there is no way (at least I could not found one) that was one of the "issues" of using library which I mentioned. They are not included in the standard `ng build`

**Comment by michal-szostak** [comment]
> I have a general question, why there are two types of NotificationComponent: DynamicNC and GeneralNC. Dynamic seems to be used in the app, while I don't see the binding to the General. However, I'm probably missing something. Could you explain this architecture?

Sure. The `DynamicNC` is a dynamic component which instantiates concrete notification component based on passed `notification` input.

The coupling is loose on purpose and mappings between types and components is done in providers (see `notifications.module.ts` there are providers like: `{provide: NotificationClazz, useValue: {name: 'GeneralNotificationComponent', component: GeneralNotificationComponent}, multi: true},`)

This kind of architecture allows extending notifications (providing custom notifications with custom data) outside of the library. For example one could create custom notification component which shows progress bar of loading some resources, or something.

File: s4e-web/package.json (hunk 4 above)
Comment (jswk) [reply s4e-web/package.json:8]: Understood, thanks for clarification

File: s4e-web/projects/notifications/src/lib/components/general-notification/general-notification.component.html (hunk 1 above)
Comment (michal-szostak) [reply s4e-web/projects/notifications/src/lib/components/general-notification/general-notification.component.html:7]: They are for @abacz . I opted out of using caption and extended elements, but this is just my humble opinion. Maybe @abacz will decide that notifications should have captions, etc. That's why I left it here, as a hint.

File: s4e-web/projects/notifications/src/lib/state/notification.service.spec.ts (hunk 2 above)
Comment (michal-szostak) [reply s4e-web/projects/notifications/src/lib/state/notification.service.spec.ts]: This is the test case which had to be done (it should be good now)

**Pull request merged**